expose cryptography primitives to wasm

MGibson1 opened the pull request:
## 📔 Objective

This is a stop-gap to allow encryption changes to be developed only in the rust SDK and used from the javascript clients.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

MGibson1 committed:
expose cryptography primitives to wasm

This is a stop-gap to allow encryption changes to be developed only in the rust SDK and used from the javascript clients.

quexten commented on crates/bitwarden-wasm-internal/src/client.rs:
```diff
@@ -27,6 +27,56 @@ fn convert_level(level: LogLevel) -> Level {
+pub struct BitwardenPure;
+
+#[wasm_bindgen]
+impl BitwardenPure {
```
Since this is pub, could we get some docs on it so it's clear that teams should not be consuming this directly?

quexten commented on crates/bitwarden-wasm-internal/src/client.rs:
```diff
@@ -27,6 +27,56 @@ fn convert_level(level: LogLevel) -> Level {
+        msg
+    }
+
+    pub fn throw(msg: String) -> Result<(), TestError> {
```
```suggestion
```

quexten commented on crates/bitwarden-wasm-internal/src/client.rs:
```diff
@@ -27,6 +27,56 @@ fn convert_level(level: LogLevel) -> Level {
+        env!("SDK_VERSION").to_owned()
+    }
+
+    pub fn echo(msg: String) -> String {
```
```suggestion
```

Seems like left over debugging code?

MGibson1 replied:
No I included those because they are present in the `BitwardenClient`, but more appropriately part of a pure function library.

quexten committed:
Cargo fmt

quexten committed:
Cleanup

quexten commented on crates/bitwarden-wasm-internal/src/client.rs:
```diff
@@ -27,6 +31,56 @@ fn convert_level(level: LogLevel) -> Level {
+        pure_crypto::symmetric_decrypt(enc_string, key_b64)
+    }
+
+    pub fn symmetric_decrypt_to_bytes(
```
I think we want an additional function here that decrypts the enc buffers (attachments)

MGibson1 committed:
Add `EncByteArray` handling

MGibson1 committed:
segment pure function namespace by purpose

MGibson1 commented:
I decided to take your recommendation on the helper functions and separated the `BitwardenPure` namespace into different purposes.

this left `bitwarden_pure` as a reasonable place to put the required startup code to link panics and errors to console.

MGibson1 committed:
various formatters

quexten commented on crates/bitwarden-wasm-internal/src/crypto.rs:
```diff
@@ -8,8 +8,113 @@ use bitwarden_core::{
+    ///
+    /// This method is intended for use in the javascript clients at the EncryptService layer and
+    /// should not be used elsewhere.
+    pub fn symmetric_decrypt_to_bytes(
```
Nice, change look good. One more thing: How would cases where the encrypted version is an encstring but the decrypted version is a buffer be handeled here? So f.e a userkey is an encstring, but would not necessarily be valid utf8. Would the caller need to convert their encstring to the buffer representation first? Do we want a separate function here?

quexten commented on crates/bitwarden-wasm-internal/src/crypto.rs:
```diff
@@ -8,8 +8,113 @@ use bitwarden_core::{
+    ///
+    /// This method is intended for use in the javascript clients at the EncryptService layer and
+    /// should not be used elsewhere.
+    pub fn symmetric_encrypt(plain: String, key_b64: String) -> Result<String, PureCryptoError> {
```
Optional: encrypt_bytes? (So a userkey can be encrypted)

Hinton commented:
Requesting a review from platform since this is named `PureClient` implying usages beyond crypto and therefore platform ownership.

Hinton commented on crates/bitwarden-wasm-internal/src/crypto.rs:
This diverges from the established convention of just having the wasm-internal crate be a wasm wrapper.

coroiu replied:
I agree with Oscar here, I suggest creating a new crate `bitwarden-pure-crypto` or moving `PureCrypto` to the existing `bitwarden-crypto`. You can also use conditional compilation for the functions if you only want these functions to be available in wasm

MGibson1 replied:
I don't understand. How is this not a wrapper? The only thing it does is convert a few types because we don't necessarily want to write that conversion in javascript.

A crate doing that seems entirely out of proportion to what it would be handling. In that case we'd have a crate that calls a method that literally just converts a string to a byte array to call the actual cryptographic implementation. What's more, we would explicitly not want to use that crate for anything other than WASM bidings. The `bitwarden-crypto` crate also shouldn't be polluted with the weak typings necessary for javascript interoperability.

Are you disagreeing to this being the crypto file?

coroiu replied:
It uses multiple types to create a utility function which might be useful outside of WASM. I just think that we shouldn't default to putting these in the wasm crate and instead opt for putting them in a common place such as the `bitwarden-crypto` crate.

> The bitwarden-crypto crate also shouldn't be polluted with the weak typings necessary for javascript interoperability.

We're actually actively trying to get rid of as much as possible of the wasm crate and instead try to keep everything related to the same feature closer together.

quexten replied:
> It uses multiple types to create a utility function which might be useful outside of WASM 

But the explicit purpose of the functions introduced in this PR is a stop-gap for javascript clients that can't use the SDK yet for encrypting / decrypting all domains, but need support for e.g. encstrings for new ciphertypes not available in typescript.
It really has no use outside of WASM/typescript; the intent is to remove it again as soon as *all* domains have been moved.

coroiu replied:
Ok so you just want to expose the existing `bitwarden-crypto` crate to WASM? Just add the methods to the existing `CryptoClient` :) If you want to separate it from the existing stuff maybe create a sub-client called `TemporaryCryptoClient`, `RawCryptoClient`, `SubtleCryptoClient` or maybe even `PureCryptoClient`. I'm not gonna block the addition of a separate `PureCrypto` struct (I'm adding a similar standalone `IpcManager`) but if you go that route then I don't think it should be calling `setup_once()` everywhere (see comment further down)

MGibson1 commented:
> Requesting a review from platform since this is named PureClient implying usages beyond crypto and therefore platform ownership.

There's no `PureClient`, just the crypto one for now. I do think that the [helper functions present in the `BitwardenClient` ](https://github.com/bitwarden/sdk-internal/blob/b24c9736c10b8b56857b1faded309faf363d5e76/crates/bitwarden-wasm-internal/src/client.rs#L62-L74) make more sense to be part of exposed pure functions.

MGibson1 committed:
Prefer feature impl to wrapper complexity

coroiu commented on crates/bitwarden-wasm-internal/src/crypto.rs:
```diff
@@ -8,8 +8,113 @@ use bitwarden_core::{
+pub enum PureCryptoError {
+    #[error("Cryptography error, {0}")]
+    Crypto(#[from] bitwarden_crypto::CryptoError),
+}
```
suggestion: just add `bitwarden_error` to `bitwarden_crypto::CryptoError` and return that directly

MGibson1 replied:
This is not an error we want to expose long term. I don't think there's a need to compile this macro's expansion for all crates depending on crypto

Hinton replied:
We've already added `bitwarden_error` to `CryptoError` since we now explicitly expose the underlying errors instead of the grouped error struct.

I don't see a reason to keep the `PureCryptoError` any longer.

coroiu commented on crates/bitwarden-wasm-internal/src/client.rs:
```diff
@@ -27,6 +27,19 @@ fn convert_level(level: LogLevel) -> Level {
+        if let Err(_e) = console_log::init_with_level(log_level) {
+            set_max_level(log_level.to_level_filter())
+        }
+    }
```
issue: Can we move/remove this? We want to eventually add more control over the initialization process and this will make it much harder to coordinate that. I don't think calling `symmetric_decrypt()` should implicitly mean setting the logging level to `Info`. For context I don't think setting it in `BitwardenClient::new` is the right way either. I like what you did with the `SdkLoadService` and I think we should try to move the `init` there instead. I also think that consumers of the SDK in the clients might need to depend on the SDK initialization somehow and not just assume that WASM has been initialized by the time their function runs

MGibson1 replied:
This is kind of tricky.

The issue is that these methods are otherwise pure and don't need to depend on any previous calls or state. In the case of the panic hook, I think what is here is fine. The documentation specifically calls out subsequent invocations doing nothing.

All that can be done with your suggesting is taking an optional `LogLevel` in the signature of the pure functions and passing it along here. I didn't think that was worthwhile.

coroiu commented:
@MGibson1 Just for some context, why not just put these functions in their own `Client` and expose them through the `BitwardenClient`? They could still be "pure" in the sense that they don't use any state and it would follow existing patterns. Is it because the call would become `async` in js? Or is it because you think this has static typing guarantees about not needing a logged in user? We are working together with Auth on solving that btw

MGibson1 commented:
The logged in user guarantees are worrying in general, but aren't really motivations here.

These aren't in the `BitwardenClient` because they don't need the added complexity associated with initializing a user-bound client. The intent is to temporarily expose cryptographic primitives to the JS clients while those clients are still in control of keys. The situation is distinct from the mobile clients, because the SDK is in control of user keys in those applications.

As a background, a new `EncString` type is coming that will be implemented only in the SDK. However, we want to meter transition to this type slowly and allow for clients to decrypt these values long before we actually turn on their use. These methods will be used to tie into the SDKs `EncString` processing while allowing the JS clients to maintain control of user keys. By exposing primitive methods that don't need any external state or setup, we can enable JS clients to handle the new crypto type, and use a feature flag to turn on creating them in JS.

MGibson1 committed:
Revert "Prefer feature impl to wrapper complexity"

This reverts commit 10634581d6c6eb8fb7d47ad9357f85d830a3843f.

MGibson1 committed:
Cleanup rebase. pure initialization is handled during module init

MGibson1 committed:
Fix format

Hinton commented on crates/bitwarden-wasm-internal/src/crypto.rs:
**suggestion:** We can move all of the pure-crypto content into a separate file, which will make it easier to remove later. And add a file level comment with disclaimers to not re-use this pattern which should make it explicitly clear this is not a sanctioned pattern.

Problems with this pattern:
- Logic in bitwarden-wasm-internal
- untested code.

Hinton commented on crates/bitwarden-wasm-internal/Cargo.toml:
```diff
@@ -26,6 +26,7 @@ console_log = { version = "1.0.0", features = ["color"] }
 js-sys = "0.3.68"
 log = "0.4.20"
 serde_json = ">=1.0.96, <2.0"
+thiserror.workspace = true
```
**nit:** We generally use this syntax everywhere else.
```suggestion
thiserror = { workspace = true }
```

Hinton reviewed (changes requested):
I want a more clear disclaimer that this pattern is forbidden from propagating to make it clear to not copy this behaviour in the future.

MGibson1 committed:
Prefer direct errors

CryptoError has since been decorated with the `bitwarden_error` and can now be directly returned by wasm

MGibson1 committed:
Move to unique module and test

Hinton commented on crates/bitwarden-wasm-internal/src/pure_crypto.rs:
```diff
@@ -0,0 +1,116 @@
+
```
Linting seems to be failing.

```suggestion
```

MGibson1 committed:
Update crates/bitwarden-wasm-internal/src/pure_crypto.rs

Co-authored-by: Oscar Hinton <Hinton@users.noreply.github.com>

MGibson1 committed:
`cargo +"nightly-2025-01-01" fmt` :robot:

MGibson1 committed:
`cargo clippy` :robot:
